tools/build_variables.bzl

```diff
@@ -531,7 +532,6 @@ libtorch_python_distributed_sources = [
     "torch/csrc/distributed/rpc/python_functions.cpp",
     "torch/csrc/distributed/rpc/python_rpc_handler.cpp",
     "torch/csrc/distributed/rpc/request_callback_impl.cpp",
-    "torch/csrc/distributed/rpc/tensorpipe_agent.cpp",
```

Comment:
If we're moving `tensorpipe_agent.cpp` to libtorch then we should probably do the same with `process_group_agent.cpp` (a few lines above, GitHub won't let me comment on it), which is the other (default) RPC agent. And at this point I am wondering whether there is any specific reason for all the agents to only be included in libtorch_pytorch. @osalpekar @pritamdamania @mrshenli may know.

Reply:
Looking at the link failures more attentively, perhaps the reason to only add the agents to libtorch_python is that they must store an instance of RequestCallback, which in turn must be Python-aware. If that's the case I am not sure we can fix it without refactoring RequestCallback (which if I remember correctly is on our roadmap for v1.7, but that won't help for v1.6).

Reply:
> Looking at the link failures more attentively, perhaps the reason to only add the agents to libtorch_python is that they must store an instance of RequestCallback, which in turn must be Python-aware. If that's the case I am not sure we can fix it without refactoring RequestCallback (which if I remember correctly is on our roadmap for v1.7, but that won't help for v1.6).
According to https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/rpc/request_callback.h, `RequestCallback` is an abstract class, so it is not Python aware. And it doesn't look like any parts of `libc10d` are actually Python aware and therefore could be moved to libtorch_cpu/libtorch_cuda.

Reply:
Yes the RequestCallback class is indeed abstract, but the TensorPipe agent must then create an instance of RequestCallbackImpl which is its concrete subclass:
https://github.com/pytorch/pytorch/blob/df252c059c3ea4c04f8c39c3b4c25a3f4572b0b7/torch/csrc/distributed/rpc/tensorpipe_agent.cpp#L220
And RequestCallbackImpl starts pulling in Python stuff.